### PR TITLE
Add ServiceMonitor for metrics Service

### PR DESCRIPTION
**Pull request by stone-z**
Towards https://github.com/giantswarm/giantswarm/issues/18354

### Checklist

- [x] Update changelog in CHANGELOG.md.
- [ ] Make sure `values.yaml` and `values.schema.json` are valid.

### CODE REVIEW

**Comment by stone-z** [comment]
@QuentinBisson it looks like the [kyverno ServiceMonitor policy ](https://github.com/giantswarm/kyverno-policies/blob/main/policies/shared/ServiceMonitor.yaml) unintentionally drops fields. The ServiceMonitor in this PR works as expected if I remove the policy before applying it:

```yaml
$ kubectl get servicemonitor -n giantswarm kyverno -o yaml
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  annotations:
    meta.helm.sh/release-name: kyverno
    meta.helm.sh/release-namespace: giantswarm
  creationTimestamp: "2021-08-20T19:04:29Z"
  generation: 1
  labels:
    app.kubernetes.io/component: kyverno
    app.kubernetes.io/instance: kyverno
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: kyverno
    app.kubernetes.io/part-of: kyverno
    app.kubernetes.io/version: 0.3.0-6d60f8dace82b4d2ee50e583c9e0716e900a31ad
    helm.sh/chart: kyverno-0.3.0-6d60f8dace82b4d2ee50e583c9e0716e900a31ad
  name: kyverno
  namespace: giantswarm
  resourceVersion: "241065732"
  selfLink: /apis/monitoring.coreos.com/v1/namespaces/giantswarm/servicemonitors/kyverno
  uid: e6221044-e53f-4a36-bed1-90136e531e85
spec:
  endpoints:
  - honorLabels: true
    path: /metrics
    port: metrics-port
    relabelings:
    - action: keep
      regex: kyverno-svc-metrics;8000
      sourceLabels:
      - __meta_kubernetes_service_name
      - __meta_kubernetes_pod_container_port_number
    - sourceLabels:
      - __meta_kubernetes_endpoints_label_app_kubernetes_io_name
      targetLabel: app
  selector:
    matchLabels:
      app.kubernetes.io/instance: kyverno
      app.kubernetes.io/name: kyverno
  targetLabels:
  - __meta_kubernetes_service_name
  - __meta_kubernetes_pod_container_port_number
```

But if the policy is in place, the `port` gets stripped along with other info:

```yaml
$ kubectl get servicemonitor -n giantswarm kyverno -o yaml
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  annotations:
    meta.helm.sh/release-name: kyverno
    meta.helm.sh/release-namespace: giantswarm
    policies.kyverno.io/patches: |
      configure-service-monitor-labelling-schema.configure-service-monitor-labelling-schema.kyverno.io: removed
        /spec/endpoints/0/port
  creationTimestamp: "2021-08-20T18:59:51Z"
  generation: 1
  labels:
    app.kubernetes.io/component: kyverno
    app.kubernetes.io/instance: kyverno
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: kyverno
    app.kubernetes.io/part-of: kyverno
    app.kubernetes.io/version: 0.3.0-6d60f8dace82b4d2ee50e583c9e0716e900a31ad
    helm.sh/chart: kyverno-0.3.0-6d60f8dace82b4d2ee50e583c9e0716e900a31ad
  name: kyverno
  namespace: giantswarm
  resourceVersion: "241062898"
  selfLink: /apis/monitoring.coreos.com/v1/namespaces/giantswarm/servicemonitors/kyverno
  uid: 6c99cea0-54de-4731-b94b-07561e92cfe0
spec:
  endpoints:
  - relabelings:
    - replacement: gauss
      targetLabel: cluster_id
    - replacement: management_cluster
      targetLabel: cluster_type
    - replacement: aws
      targetLabel: provider
    - replacement: gauss
      targetLabel: installation
    - sourceLabels:
      - __meta_kubernetes_namespace
      targetLabel: namespace
    - sourceLabels:
      - __meta_kubernetes_pod_name
      targetLabel: pod
    - sourceLabels:
      - __meta_kubernetes_pod_container_name
      targetLabel: container
    - sourceLabels:
      - __meta_kubernetes_pod_node_name
      targetLabel: node
    - sourceLabels:
      - __meta_kubernetes_node_label_role
      targetLabel: role
    - regex: ""
      replacement: worker
      sourceLabels:
      - __meta_kubernetes_node_label_role
      targetLabel: role
  selector:
    matchLabels:
      app.kubernetes.io/instance: kyverno
      app.kubernetes.io/name: kyverno
  targetLabels:
  - __meta_kubernetes_service_name
  - __meta_kubernetes_pod_container_port_number
```

**Comment by QuentinBisson** [comment]
Yes we have a story to fix it in @giantswarm/team-atlas  but I'm off for 2 weeks. I think we need to change the policy to be the same as:
https://github.com/giantswarm/kyverno-policies/blob/d462806b9f31d4b5faa17ada677334959eba72f6/policies/shared/Silence.yaml#L13

This ensure this only adds items and not replace the whole CR. If you have time, could you try to update the policy and test the 2 PRs together ?

**Comment by MarcelMue** [comment]
@stone-z We can pair today to fix this if you have time.

File: helm/kyverno/templates/servicemonitor.yaml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  selector:
+    matchLabels: {{ include "kyverno.matchLabels" . | nindent 6 }}
+  endpoints:
+  - honorLabels: true
+    path: /metrics
+    port: metrics-port
+    relabelings:
+    - sourceLabels: [__meta_kubernetes_service_name, __meta_kubernetes_pod_container_port_number]
+      regex: "{{ template "kyverno.serviceName" . }}-metrics;{{ .Values.metricsService.port }}"
+      action: keep
```
Comment (mig4) [inline helm/kyverno/templates/servicemonitor.yaml:17]: Why is this filter needed? I.e. I'd have thought the selector and `endpoint.port` should be enough to narrow down what to scrape, no?

Comment (stone-z) [reply helm/kyverno/templates/servicemonitor.yaml:17]: Sure I can try it with only the port. This was just the rule as it was originally created for PMO

**Comment by stone-z** [comment]
@MarcelMue I started fixing the policy on Friday ([branch](https://github.com/giantswarm/kyverno-policies/tree/fix-servicemonitor)), but `kyverno-policies` tests hang forever (locally and in CI). We can take a look today

**Comment by stone-z** [comment]
In looking at the `kyverno` policy mutating ServiceMonitors we might have found an upstream bug, but will deal with that mutation separately. For now, the policy has been removed.

**Review by MarcelMue — APPROVED** [review]

**Pull request merged**